### ISSUE TITLE
Termination of background tasks when user cancels the project setup activity

### ISSUE THREAD

**Issue by susmina94** [hotosm/fmtm#912]
**Describe the bug**
I created a project, had some issues with splitting the task (which was the platform not supporting the file in projected coordinate system), I couldn't proceed ahead so I just refreshed the tab and started creating new projects one after another. 

After sometime, the platform didn't respond. 
When I asked the devs, they mentioned the previous incomplete project setups kept the background tasks running even though I cancelled it or refreshed the tab.

The reason being all the workers were occupied with the previous project background tasks.
They had to restart the system manually and then it was back to its working state.

**Expected behavior**
The first behavior would be :

1. If the user cancels the project setup, it should stop all the background tasks related to it so that the workers are not loaded unnecessarily
2. If there is any error in input, validations and proper error messages should be displayed for the users to try again.

**Screenshots**
![image](https://github.com/hotosm/fmtm/assets/108750444/01fc1cf5-ca55-43b3-b7b7-a429b56990b9)

**Comment by spwoodcock** [comment]
There are two issues to consider here:

### Event Loop Blocking

- Running as async function within the background task will run it in the asyncio event loop.
- If there are no more workers to handle the request (set to 4), then FastAPI will hang.

More context here: https://stackoverflow.com/questions/67599119/fastapi-asynchronous-background-tasks-blocks-other-requests

### Cancelling Jobs

- We track background jobs in our DB (background_tasks table).
- Their status is updated when complete.
- BackgroundTasks cannot be easily stopped though. 
- Ideally we need a way to collect running background tasks associated with an endpoint (tagged with project ID), then cancel existing jobs if the user cancels, goes back, refreshes, etc.

### Possible Solution

We discussed using celery before & agreed it's too heavyweight. 

I agree - scalability is also a pain, so using the database to manage tasks is a better solution IMO.

[apscheduler](https://github.com/agronholm/apscheduler) looks like a good tool for our use case, to replace the simple BackgroundTasks.

It allows for tasks to be cancelled, paused, resumed. It could probably be implemented ourselves, but this is simple.

It can be set up with FastAPI asyncio loop:

```python
from apscheduler.schedulers.asyncio import AsyncIOScheduler

sched = AsyncIOScheduler()
sched.start()
```

Then jobs can be spawned from called endpoints with an ID associated (note no async).

```python
def job():
    sched.add_job(download_tiles, "date", run_date=datetime.now(), args=[], id=<unique_id>
    # to stop: sched.remove_job(<unique_id>)
```

**Comment by spwoodcock** [comment]
Additional context for event loop blocking:

```
In summary most tasks generally use the database and are limited by I/O, so using async def is the way to go.

When running CPU limited tasks (i.e. tasks that generally run for a long time), or tasks that ultimately call synchronous code underneath, then using async def can actually block the event loop, making the server hang. In these cases then a standard def is best. An example of this would be task splitting that is quite CPU intensive, or the basemap downloads, which are actually threaded underneath and can take a long time to run.
```

**Comment by spwoodcock** [comment]
## Using apscheduler

@nrjadkry has done some great research into this.

Apscheduler is actually quite limited:
- It creates a db table with: id, next_run_time, and task_state.
- Scheduled jobs can be cancelled before they run.
- Jobs that are currently running cannot be cancelled (it just cancelled via the db, nothing special with asyncio task cancellation).

**So apscheduler is probably not suitable for our needs.**

## Original Requirements

### Not blocking the event loop when we run create_project

This can be solved by using a def for now instead of async def and it will run in another thread.

### Cancelling running create_project

This is lower priority for now in my opinion.
We will keep this issue open to address it in the future.

In summary:

- This is partly about trust. We should hope that a user won't intentionally spam the project_create endpoint and hang the server.
- When the user submits the form to create a project, the form should be cleared so they cannot accidentally submit the same job twice.
- This can potentially be handled in the future by using `asyncio.create_task` instead.
- A name can be assigned to the asyncio job.
- Then the task can be cancelled either by an endpoint call, or when project_create is triggered twice in a row (previous is still running).
- Ref: https://stackoverflow.com/questions/64764679/how-to-cancel-asyncio-tasks-by-name

**Comment by spwoodcock** [comment]
Partially fixable by using a task queue #1124
Pending tasks can be cancelled.
Tasks that have commenced cannot be easily cancelled.

**Note**: something that will improve this is the QR code generation will likely be done dynamically going forward, instead of all being pre-generated #1077 (so the project creation time will be reduced)

**Comment by nischalstha9** [comment]
# OS Level Hack with _psutil_ and _multiprocessing_

Here I have presented an hacky method where I have tracked the PID of the background tasks using **multiprocessing** and **psutil** library. Here I haven't used BackgroundTask from FastAPI though.

Wondering how much this hack is acceptable. 🤔


```python
from fastapi import FastAPI
import time, psutil, signal, os
from multiprocessing import Process

app = FastAPI()

def kill_proc_tree(pid, sig=signal.SIGTERM, include_parent=True, timeout=None, on_terminate=None):
    """Kill a process tree (including grandchildren) with signal
    "sig" and return a (gone, still_alive) tuple.
    "on_terminate", if specified, is a callback function which is
    called as soon as a child terminates.
    """
    assert pid != os.getpid(), "won't kill myself"
    parent = psutil.Process(pid)
    children = parent.children(recursive=True)
    if include_parent:
        children.append(parent)
    for p in children:
        try:
            p.send_signal(sig)
        except psutil.NoSuchProcess:
            pass
    gone, alive = psutil.wait_procs(children, timeout=timeout, callback=on_terminate)
    return (gone, alive)


def long_task(dur: int, message=""):
    time.sleep(dur)
    print(message, "==COMPLETE==")

@app.get("/bg-task/")
async def start_bg_task():
    p = Process(target=long_task, args=(1000, "PROCESS"))
    p.start()
    pid = p.pid
    return {"message": "Task Queued", "task_id":pid}

@app.get("/kill/{pid}/")
async def kill(pid:int):
    kill_proc_tree(pid)
    return {"message": "Task Exit", "task_id":1}
```

**Comment by spwoodcock** [comment]
Nice idea!

Async python works a but differently though.
Functions are coroutines (wrapped functions) that are passed to the event loop.

In this case the coroutine is not the OS level process that can be killed, but the event loop is.

So first of all we don't have a PID to identify the task & second we can't easy kill the coroutine without stopping the entire process (which is running the FastAPI server in our case).

There may still be a solution though, as I am by no means an expert in Python async 😄

**Comment by nischalstha9** [comment]
I haven't use FastAPI BackgroundTask here. Rather I have used `from multiprocessing import Process`.
With this I am able to grab the PID and kill it. I have only verified it with `ps aux` command though.

```python
p = Process(target=long_task, args=(1000, "PROCESS"))
p.start()
pid = p.pid
```

Also I changed the long_task to async and it is killed when ran with `Process`
```python
async def long_task(dur: int, message=""):
    await asyncio.sleep(dur)
    print(message, "==COMPLETE==")
```

I have obtained following result:
<img width="1440" alt="image" src="https://github.com/hotosm/fmtm/assets/39024181/73840fa9-cf6f-468e-bce5-fd0b91db6083">

**Comment by spwoodcock** [comment]
The thing is that you are running these processes **inside** a multiprocessing process, which defeats the point of async / threading. 

Processes created with multiprocessing have their own memory space and can utilise multiple CPU cores. They also have PIDs and are killable.

Processes spawned via threading / async are not like the above. They can only run on a single core due to the GIL and each task does not have a separate PID that can be killed.

If you can make a workaround somehow not using multiprocessing (which I'm not sure is possible but feel free to test), then that's the only way.

**Comment by spwoodcock** [comment]
Also I should have noted that we don't want to use multiprocessing with FastAPI and async as it's very resource inefficient. It's fully synchronous then and would easily overload the server if we start spawning processes using multiprocessing.

**Comment by spwoodcock** [comment]
Project creation now takes less than 10s with Entities.

Closing this as unrealistic to cancel the creation & is technically difficult too.
